File: ios/VCInvestmentDiscovery/ViewModels/AlertCenterViewModel.swift
import Foundation
import Combine

@MainActor
class AlertCenterViewModel: ObservableObject {
    @Published var alerts: [Alert] = []
    @Published var filteredAlerts: [Alert] = []
    @Published var selectedFilter: AlertFilter = .all
    
    private let dataService: DataService
    private var cancellables = Set<AnyCancellable>()
    
    init(dataService: DataService) {
        self.dataService = dataService
        setupBindings()
        loadAlerts()
    }
    
    private func setupBindings() {
        dataService.$alerts
            .sink { [weak self] alerts in
                self?.alerts = alerts
                self?.applyFilter(self?.selectedFilter ?? .all)
            }
            .store(in: &cancellables)
        
        $selectedFilter
            .sink { [weak self] filter in
                self?.applyFilter(filter)
            }
            .store(in: &cancellables)
    }
    
    func loadAlerts() {
        Task {
            await dataService.fetchAlerts()
        }
    }
    
    func refreshAlerts() {
        loadAlerts()
    }
    
    func markAlertAsRead(_ alertId: UUID) {
        Task {
            await dataService.markAlertAsRead(alertId)
            if let index = alerts.firstIndex(where: { $0.id == alertId }) {
                alerts[index].isRead = true
            }
            applyFilter(selectedFilter)
        }
    }
    
    func deleteAlert(_ alertId: UUID) {
        Task {
            await dataService.deleteAlert(alertId)
            alerts.removeAll { $0.id == alertId }
            applyFilter(selectedFilter)
        }
    }
    
    func applyFilter(_ filter: AlertFilter) {
        selectedFilter = filter
        filteredAlerts = alerts.filter { alert in
            switch filter {
            case .all:
                return true
            case .unread:
                return !alert.isRead
            case .highPriority:
                return alert.priority == .high || alert.priority == .urgent
            }
        }
    }
    
    func sortAlerts(by criteria: AlertSortCriteria) {
        filteredAlerts.sort { alert1, alert2 in
            switch criteria {
            case .date:
                return alert1.createdAt > alert2.createdAt
            case .priority:
                return alert1.priority.rawValue > alert2.priority.rawValue
            case .type:
                return alert1.type.rawValue < alert2.type.rawValue
            }
        }
    }
}

File: android/app/src/main/java/com/vcinvestmentdiscovery/viewmodels/AlertCenterViewModel.kt
package com.vcinvestmentdiscovery.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject
import com.vcinvestmentdiscovery.data.repository.AlertRepository
import com.vcinvestmentdiscovery.data.models.*
import com.vcinvestmentdiscovery.util.Result
import com.vcinvestmentdiscovery.ui.alerts.AlertCenterUiState
import java.util.UUID

@HiltViewModel
class AlertCenterViewModel @Inject constructor(
    private val repository: AlertRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(AlertCenterUiState())
    val uiState: StateFlow<AlertCenterUiState> = _uiState.asStateFlow()

    init {
        loadAlerts()
    }

    fun loadAlerts() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true)
            when (val result = repository.getAlerts()) {
                is Result.Success -> {
                    _uiState.value = _uiState.value.copy(
                        alerts = result.data,
                        filteredAlerts = applyFilter(result.data, _uiState.value.selectedFilter),
                        isLoading = false
                    )
                }
                is Result.Error -> {
                    _uiState.value = _uiState.value.copy(
                        error = result.exception.message ?: "Unknown error occurred",
                        isLoading = false
                    )
                }
            }
        }
    }

    fun refreshAlerts() {
        loadAlerts()
    }

    fun markAlertAsRead(alertId: UUID) {
        viewModelScope.launch {
            when (val result = repository.markAlertAsRead(alertId)) {
                is Result.Success -> {
                    val updatedAlerts = _uiState.value.alerts.map { alert ->
                        if (alert.id == alertId) alert.copy(isRead = true) else alert
                    }
                    _uiState.value = _uiState.value.copy(
                        alerts = updatedAlerts,
                        filteredAlerts = applyFilter(updatedAlerts, _uiState.value.selectedFilter)
                    )
                }
                is Result.Error -> {
                    _uiState.value = _uiState.value.copy(
                        error = result.exception.message ?: "Failed to mark alert as read"
                    )
                }
            }
        }
    }

    fun deleteAlert(alertId: UUID) {
        viewModelScope.launch {
            when (val result = repository.deleteAlert(alertId)) {
                is Result.Success -> {
                    val updatedAlerts = _uiState.value.alerts.filter { it.id != alertId }
                    _uiState.value = _uiState.value.copy(
                        alerts = updatedAlerts,
                        filteredAlerts = applyFilter(updatedAlerts, _uiState.value.selectedFilter)
                    )
                }
                is Result.Error -> {
                    _uiState.value = _uiState.value.copy(
                        error = result.exception.message ?: "Failed to delete alert"
                    )
                }
            }
        }
    }

    fun applyFilter(filter: AlertFilter) {
        _uiState.value = _uiState.value.copy(
            selectedFilter = filter,
            filteredAlerts = applyFilter(_uiState.value.alerts, filter)
        )
    }

    fun sortAlerts(criteria: AlertSortCriteria) {
        val sortedAlerts = when (criteria) {
            AlertSortCriteria.DATE_DESC -> _uiState.value.filteredAlerts.sortedByDescending { it.createdAt }
            AlertSortCriteria.DATE_ASC -> _uiState.value.filteredAlerts.sortedBy { it.createdAt }
            AlertSortCriteria.PRIORITY_DESC -> _uiState.value.filteredAlerts.sortedByDescending { it.priority }
            AlertSortCriteria.PRIORITY_ASC -> _uiState.value.filteredAlerts.sortedBy { it.priority }
        }
        _uiState.value = _uiState.value.copy(filteredAlerts = sortedAlerts)
    }

    private fun applyFilter(alerts: List<Alert>, filter: AlertFilter): List<Alert> {
        return alerts.filter { alert ->
            (filter.type == null || alert.type == filter.type) &&
            (filter.priority == null || alert.priority == filter.priority) &&
            (filter.isRead == null || alert.isRead == filter.isRead)
        }
    }
}